les/server.go

```diff
@@ -130,6 +147,13 @@ func NewLesServer(node *node.Node, e *eth.Ethereum, config *eth.Config) (*LesSer
 	node.RegisterAPIs(srv.APIs())
 	node.RegisterLifecycle(srv)
 
+	// disconnect all peers at nsm shutdown
+	ns.SubscribeField(clientPeerField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if state.Equals(serverSetup.OfflineFlag()) && oldValue != nil {
+			oldValue.(*clientPeer).Peer.Disconnect(p2p.DiscRequested)
```

Comment:
I think we should have a function which will:

- Delete the peer from all components: e.g. clientpool, fetcher, e.g.
- Disconnect the network connection
- Drop it from the NSM

Reply:
I don't think so, see above. The peer logic is much simpler now and the disconnection always happens the same way. First in the p2p layer, then in the server handler, then in the NSM, then everywhere else. We don't need to call each component directly anymore. If you want to disconnect, always do `Peer.Disconnect`, if you want to do something with connecting/disconnecting peers then subscribe to NSM.